### ISSUE TITLE
Using `setFileDetectorMethod` to allow for selenium file uploads

### ISSUE THREAD

**Issue by jstockwin** [ropensci/RSelenium#130]
### Operating System
Ubuntu 17
### Selenium Server version (selenium-server-standalone-3.0.1.jar etc.)
Whatever is used by Travis CI
### Browser version (firefox 50.1.0, chrome 54.0.2840.100 (64-bit) etc.)
Firefox 45, although will be an issue with all browsers

### Issue

I am trying to use RSelenium to test an R Shiny app which requires a file to be uploaded. Selenium includes a file upload feature - when running tests locally, you can simply use `sendKeys` to type in the local path of the file. However, when running externally (e.g. I am using Travis, along with a Sauce Connect Tunnel through to SauceLabs) you need to use the `setFileDetector` method ([source](https://saucelabs.com/resources/articles/best-practices-tips-selenium-file-upload)). 

Is this possible in RSelenium? If not, would it be possible to implement (happy to try and implement it if you point me in the right direction!).

**Comment by johndharrison** [comment]
Looking at https://github.com/SeleniumHQ/selenium/pull/219/files this utilises an undocumented endpoint
`/session/:sessionId/file` in the JSONWireProtocol. A file is zipped and then base64 encoded and the resulting
string representation is POSTed to the `/session/:sessionId/file` endpoint. The issue here would be cross-platform zipping and possibly base 64 encoding. zip support in base R currently isnt great on some platforms. 

The alternative is to upload your file to SauceLabs using their API. Then the file would be available `locally`. 
The RSauceLabs package https://github.com/johndharrison/RSauceLabs implements the SauceLabs REST api though I havent looked at it foe sometime so items may have changed. See for example:
https://johndharrison.github.io/RSauceLabs/uploadFile.html and
https://johndharrison.github.io/RSauceLabs/getStoredFiles.html

**Comment by jstockwin** [comment]
@johndharrison Thanks for your quick reply.

Fair enough. I'm never really sure how closely RSelenium relates to selenium itself. 

Uploading to SauceLabs sounds like a good solution, and your package looks really handy, I'll give it a try, thanks!

Quick question, I've been looking up the sauce labs storage docs [here](https://wiki.saucelabs.com/display/DOCS/Temporary+Storage+Methods) but can't seem to find out where the files are stored? I.e. what file path to I need to use with `sendKeys` once the file is uploaded? Or do I just need the filename? (Sorry if this is an obvious question, but I figured you'd probably know this off the top of your head, whereas I am struggling to find the info).

**Comment by johndharrison** [comment]
The file should be available at `sauce-storage:myfile.txt`. If you were to upload `myfile.txt`.  I have not tested whether this is true however.

Implementing a utility function to use the `/session/:sessionId/file` endpoint would be straightforward depending on your OS. The steps would be:

1. zip your file.
2. base64 encode the zipped file. (this gives you a character string)
3. POST your base64 encoded string to `/session/:sessionId/file`
4. The selenium server should return the remote location of the file
5. Use this remote file location with sendKeys.

**Comment by jstockwin** [comment]
So, in the first instance, I've successfully uploaded the file, however when I try to use `sauce-storage` I get the following error, which I don't really know how to solve...
```
> sendKeys(remDr, pages$state2.1$selectors$incidenceDataUploadInput,
+                       "sauce-storage:PennsylvaniaH1N12009FluData.csv") 

Selenium message:unknown error: path is not absolute: sauce-storage:PennsylvaniaH1N12009FluData.csv
  (Session info: chrome=48.0.2564.97)
  (Driver info: chromedriver=2.21.371459 (36d3d07f660ff2bc1bf28a75d1cdabed0983e7c4),platform=Linux 3.13.0-83-generic x86)

Error: 	 Summary: UnknownError
 	 Detail: An unknown server-side error occurred while processing the command.
	 Further Details: run errorDetails method
```
Is there something else I need to do to enable this?


In terms of POSTing it myself, I've tried (using `POST` from `httr`):
```
POST("https://jstockwin:<access-key>@saucelabs.com/session/:<sessionID>/file", body=list(file=fileString))

```
with `<access-key>` and `<sessionID>` replaced with my access key and the session id, and where `fileString` is the base64 encoded zipped file. This just returns a 404. I've tried just setting `body=fileString`, and even just no `body` to see what happened, but everything just seems to 404. Also, I'm using a sauce connect tunnel, so have tried replacing `@saucelabs.com` with `@localhost` but that just immediately says `Couldn't connect to server` and `@localhost:4445` which times out.

Sorry, I've probably got a little out of my depth now, as I don't really know what else to try to get it working...

**Comment by johndharrison** [comment]
Something like the following should allow you to upload a file using the undocumented endpoint:

```

myUpload <- function(remDr, myfile){
  tmpfile <- tempfile(fileext = ".zip")
  # zip file
  zip(tmpfile, myfile)
  # base64 encode
  zz <- file(tmpfile, "rb")
  ar <- readBin(tmpfile, "raw", file.info(tmpfile)$size)
  encFile <- base64encode(ar)
  close(zz)
  qpath <- sprintf("%s/session/%s/file", remDr$serverURL, 
                   remDr$sessionInfo[["id"]])
  queryRD(qpath, "POST", qdata = encFile)
  # return result from server
  remDr$value
}
```

**Comment by jstockwin** [comment]
Hmm, so I've copied `queryRD` from your repo, and made it actually return the response (`res`) from the request. I get a status of 400 and 
```
> httr::content(res, simplifyVector=F)
No encoding supplied: defaulting to UTF-8.
{xml_document}
<html>
[1] <body><p>ERROR Invalid JSON in request body: 'UEsDBBQAAgAIAJ1Zx0qqebq2JgA ...
```
(The `UEsDBBQ...` is the string of the encoded file)
I can't really see anything in the original PR on their end that you linked about what form it is expecting this data to be in... I tried making `qdata = '{file: "UEs..."}'` so that it's actually JSON, but that didn't change anything.

**Comment by johndharrison** [comment]
Yes it maybe defined as:

```
myUpload <- function(remDr, myfile){
  tmpfile <- tempfile(fileext = ".zip")
  # zip file
  zip(tmpfile, myfile)
  # base64 encode
  zz <- file(tmpfile, "rb")
  ar <- readBin(tmpfile, "raw", file.info(tmpfile)$size)
  encFile <- caTools::base64encode(ar)
  close(zz)
  qpath <- sprintf("%s/session/%s/file", remDr$serverURL, 
                   remDr$sessionInfo[["id"]])
  queryRD(qpath, "POST", qdata = list(file = encFile))
  # return result from server
  remDr$value
}
```

so it would be json encoded as `{file:'sfsfasffwerfe....'}`. The endpoint unfortunately is undocumented. 
A stackoverflow post relating to PHP https://stackoverflow.com/questions/10559728/uploading-files-remotely-on-selenium-webdriver-using-php suggests the above change.

**Comment by jstockwin** [comment]
Hey, so this is really close, thanks! I think because I don't have `queryRD` properly setup locally (and don't seem to be able to get at them just from having `RSelenium` installed) I run in to a few problems, but I have now managed to get it to upload a file, and I can see that there is a `value` in the response which looks like `/tmp/.com.google.Chrome.ZHQUkn/uploadZ7ifjb/home`. 

I used `sendKeys` and things seemed to be working but now I am hitting an error (server side) saying `empty beginning of file`. My app is also built to only accept `.csv` (and similar) files, so this may also be a bit of a problem, but hopefully this is something I can work around somehow. Thanks!

**Comment by jstockwin** [comment]
Ahhh, I've got it. 

Basically that `/home` replicates my own home folder. So I need to use `/tmp/.com.google.Chrome.ZHQUkn/uploadZ7ifjb/home/jake/...../file.csv` with `sendKeys` and then it works nicely. 

Incidentally, I worked this out by actually clicking the "browse" button of my app during a (fake) test through the sauce labs web interface, which actually allowed me to browse to `/tmp/.com...` and work out what was going on. Sauce Labs impress me every time I use them!

Thanks @johndharrison for putting the time into this, much appreciated! If I can ever help with anything then let me know!
